Add hooks platform

chrisccoulson opened the pull request:
This adds a KeyData platform that permits sealing and recovery of keys
via hooks provided via a device's gadget and kernel snaps, for platforms
where we don't have a native secboot platform.

chrisccoulson commented:
One thing I've just noticed here is that I've missed some integration tests between the new platform and `secboot.KeyData`. I'll push another update with those in, although it doesn't stop anyone reviewing what's already here.

pedronis commented on hooks/platform.go:
```diff
@@ -0,0 +1,158 @@
+)
+
+const (
+	platformName = "fde-hooks-v3"
```
I wonder if for consistency with the other naming and the context of all of this we need the "fde-" of the name here?

chrisccoulson replied:
I'm not sure - I just based this on what's currently in snapd (`fde-hooks-v2`)

pedronis replied:
ah, I forgot about that. This is not unreasonable then

pedronis commented on hooks/keydata.go:
```diff
@@ -0,0 +1,284 @@
+		var encryptedKey []byte
+		encryptedKey, handle, err = keyProtector.ProtectKey(rand, symKey, nil)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect symmetric key for AEAD compat: %w", err)
```
it would be good to have a test for the error path coming from the Protector

chrisccoulson replied:
I've added a test for this now.

pedronis commented on hooks/keydata.go:
```diff
@@ -0,0 +1,284 @@
+	default:
+		ciphertext, handle, err = keyProtector.ProtectKey(rand, payload, aad)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect key: %w", err)
```
same here

chrisccoulson replied:
And here

pedronis commented on hooks/keydata.go:
```diff
@@ -0,0 +1,284 @@
+//
+// On success, this will automatically update the corresponding *[secboot.KeyData]
+// that this key data was created from using [NewKeyData].
+func (d *KeyData) SetAuthorizedSnapModels(key secboot.PrimaryKey, models ...secboot.SnapModel) error {
+	if err := d.data.Scope.SetAuthorizedSnapModels(key, d.k.Role(), models...); err != nil {
+		return err
+	}
+	return d.k.MarshalAndUpdatePlatformHandle(d)
+}
+
+// SetAuthorizedBootModes updates the boot modes that this key data is
+// bound to. This update has to be authenticated, which requires the primary
+// key that was originally supplied to [NewProtectedKey].
+//
+// On success, this will automatically update the corresponding *[secboot.KeyData]
+// that this key data was created from using [NewKeyData].
+func (d *KeyData) SetAuthorizedBootModes(key secboot.PrimaryKey, modes ...string) error {
+	if err := d.data.Scope.SetAuthorizedBootModes(key, d.k.Role(), modes...); err != nil {
+		return err
+	}
+	return d.k.MarshalAndUpdatePlatformHandle(d)
+}
```
these methods don't seem to be exercised in tests

chrisccoulson replied:
I've added some tests for both of these methods now.

pedronis commented on hooks/platform.go:
```diff
@@ -0,0 +1,158 @@
+	switch {
+	case kd.data.AEADCompat != nil:
+		symKey, err := keyRevealer.RevealKey(kd.data.Handle, kd.data.AEADCompat.EncryptedKey, nil)
+		if err != nil {
```
same for the revealer

chrisccoulson replied:
I've added tests for key revealer errors, plus tests for several other error paths.

pedronis commented on hooks/platform.go:
```diff
@@ -0,0 +1,158 @@
+		return payload, nil
+	default:
+		payload, err := keyRevealer.RevealKey(kd.data.Handle, encryptedPayload, aad)
+		if err != nil {
```
and here

chrisccoulson replied:
There's a test for this one as well.

pedronis reviewed (commented):
thank you, did a first pass

chrisccoulson commented:
I've added most tests now with the exception of tests for `KeyData.SetAuthorizedSnapModels` and `KeyData.SetAuthorizedBootModes`,which I'll do after the weekend. I've also added a bunch of extra tests for error paths in `hooksPlatform.RecoverKeys`.

I'm going to add some tests for the JSON serialization next week as well. In order to avoid non-determinism in the outputs here (because of the ECDSA signatures in `bootscope` which consume randomness), I've added an `io.Reader` for functions that consume random bytes so that unit tests can override the source of randomness with fixed sets of bytes where they need determinism, and plumbed this into `bootscope`. I've had to add an additional helper to `internal/testutil` called `BypassMaybeReadByte` to bypass the usage of go's internal `randutil.MaybeReadByte` which is used in various go crypto packages in order to make the results of APIs non-deterministic even for the same sequence of input bytes.

chrisccoulson commented:
The 1.22 tests are failing unfortunately due to change in nonce generation when creating ECDSA signatures, although it's still deterministic (well, it is with my `BypassMaybeReadByte` hack. I'm re-working them to try to keep the hardcoded byte sequences in the tests as I think it's important for detecting changes that might have backwards compatibility implications (particularly with the marshalling / unmarshalling tests - and I'd like to use this same approach for testing in other packages), but this is a spanner in the works I could have done without tbh :(

pedronis reviewed (approved):
thanks
